# Data Cleaning

## Student Performance Dataset

In this notebook, I will:

- Load the raw student performance dataset
- Inspect the dataset
- Check rows and columns
- Check data types
- Identify missing values
- Identify duplicate records
- Check invalid values
- Check inconsistent text values
- Clean the dataset
- Save the cleaned dataset separately

Important:
The original raw CSV file will never be modified.
The cleaned dataset will be saved in data/processed/.

In [4]:
import pandas as pd
import numpy as np

In [5]:
file_path = "../data/raw/student_performance_raw.csv"

df = pd.read_csv(file_path)

print("Dataset loaded successfully!")

Dataset loaded successfully!


In [6]:
df.head()

,Student_ID,Student_Name,Age,Gender,Department,Semester,Study_Hours_Per_Day,Attendance_Percentage,Previous_Semester_Marks,Assignment_Average,Internal_Test_Marks,Practical_Marks,Participation_Score,Sleep_Hours,Internet_Access,Study_Resources,Backlogs,Final_Exam_Marks,Pass_Fail,Performance_Category
0,S23AIM001,Tanushree Tiwari,20,Female,CSE (AI & ML),5,3.5,89.5,76.5,19.0,25.0,40,9.0,7.0,Yes,Textbooks,0,77.0,Pass,Very Good
1,S23AIM002,Nitin Kundu,21,Male,CSE (AI & ML),5,1,71.5,32.0,8.0,4.0,22,7.0,7.0,Yes,Online Courses,4,16.0,Fail,Needs Improvement
2,S23AIM003,Amit Laha,21,Male,CSE (AI & ML),5,5.5,78.4,82.0,20.0,26.5,45,5.0,6.0,Yes,Notes Only,2,80.0,Pass,Very Good
3,S23AIM004,Akash Thakur,21,Male,CSE (AI & ML),6,4,78.9,67.7,16.5,21.5,38,6.0,5.0,Yes,Online Courses,0,63.0,Pass,Good
4,S23AIM005,Aarav Rao,22,Male,cse aiml,6,2.94,65.4,49.0,10.0,13.0,34,6.0,5.0,Yes,Both,0,47.0,Pass,Needs Improvement


In [7]:
df.tail()

,Student_ID,Student_Name,Age,Gender,Department,Semester,Study_Hours_Per_Day,Attendance_Percentage,Previous_Semester_Marks,Assignment_Average,Internal_Test_Marks,Practical_Marks,Participation_Score,Sleep_Hours,Internet_Access,Study_Resources,Backlogs,Final_Exam_Marks,Pass_Fail,Performance_Category
2535,S24AIM188,MOUMITA BASU,19,Female,CSE (AI & ML),3,1,71,46.8,8.0,13.0,36,7.0,8.5,Yes,Both,2,28.0,Fail,Needs Improvement
2536,S24CSE145,HARSH ANSARI,19,Male,Computer Science & Engineering,4,NaN,92.9,75.0,14.5,23.0,41,7.0,8.5,Yes,Both,0,85.0,Pass,Very Good
2537,S24CSE349,RISHABH BANERJEE,20,Male,Computer Science & Engineering,4,1.5,69.2,54.0,12.5,15.0,24,2.0,7.0,Yes,Online Courses,3,32.0,Fail,Needs Improvement
2538,S24AIM150,AARAV KUNDU,19,Male,CSE (AI & ML),3,4.51,75.7,66.4,NaN,20.5,32,7.0,6.0,Yes,Online Courses,0,76.0,Pass,Very Good
2539,S24ECE184,PRATIK HALDER,20,Male,Electronics & Communication,3,2,64.1,48.0,13.5,15.0,29,6.0,4.5,Yes,Textbooks,0,46.0,Pass,Needs Improvement


In [8]:
df.shape

(2540, 20)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2540 entries, 0 to 2539
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Student_ID               2540 non-null   object 
 1   Student_Name             2540 non-null   object 
 2   Age                      2540 non-null   int64  
 3   Gender                   2540 non-null   object 
 4   Department               2540 non-null   object 
 5   Semester                 2540 non-null   int64  
 6   Study_Hours_Per_Day      2477 non-null   object 
 7   Attendance_Percentage    2465 non-null   object 
 8   Previous_Semester_Marks  2490 non-null   float64
 9   Assignment_Average       2463 non-null   float64
 10  Internal_Test_Marks      2503 non-null   float64
 11  Practical_Marks          2540 non-null   int64  
 12  Participation_Score      2430 non-null   float64
 13  Sleep_Hours              2465 non-null   float64
 14  Internet_Access         

In [10]:
df.describe(include="all")

,Student_ID,Student_Name,Age,Gender,Department,Semester,Study_Hours_Per_Day,Attendance_Percentage,Previous_Semester_Marks,Assignment_Average,Internal_Test_Marks,Practical_Marks,Participation_Score,Sleep_Hours,Internet_Access,Study_Resources,Backlogs,Final_Exam_Marks,Pass_Fail,Performance_Category
count,2540,2540,2540.000000,2540,2540,2540.000000,2477,2465,2490.000000,2463.000000,2503.000000,2540.000000,2430.000000,2465.000000,2514,2464,2540.000000,2529.000000,2529,2529
unique,2500,2060,NaN,11,27,NaN,351,490,NaN,NaN,NaN,NaN,NaN,NaN,8,12,NaN,NaN,2,5
top,S23ECE113,Isha Saha,NaN,Male,Computer Science & Engineering,NaN,2,100,NaN,NaN,NaN,NaN,NaN,NaN,Yes,Both,NaN,NaN,Pass,Good
freq,2,4,NaN,1362,643,NaN,411,44,NaN,NaN,NaN,NaN,NaN,NaN,2188,803,NaN,NaN,2296,836
mean,NaN,NaN,19.947244,NaN,NaN,3.869685,NaN,NaN,63.023655,13.302071,19.543548,33.909055,6.074897,6.709939,NaN,NaN,0.376772,60.758403,NaN,NaN
std,NaN,NaN,3.740548,NaN,NaN,1.266384,NaN,NaN,11.516778,4.122158,5.076720,6.402786,1.762139,1.202862,NaN,NaN,0.743736,19.634813,NaN,NaN
min,NaN,NaN,0.000000,NaN,NaN,2.000000,NaN,NaN,32.000000,3.000000,4.000000,2.000000,1.000000,0.000000,NaN,NaN,-1.000000,-7.000000,NaN,NaN
25%,NaN,NaN,19.000000,NaN,NaN,3.000000,NaN,NaN,55.125000,11.000000,16.500000,30.000000,5.000000,6.000000,NaN,NaN,0.000000,50.000000,NaN,NaN
50%,NaN,NaN,20.000000,NaN,NaN,4.000000,NaN,NaN,62.900000,13.000000,19.500000,34.000000,6.000000,6.500000,NaN,NaN,0.000000,60.000000,NaN,NaN
75%,NaN,NaN,21.000000,NaN,NaN,5.000000,NaN,NaN,70.800000,15.000000,22.500000,38.000000,7.000000,7.500000,NaN,NaN,1.000000,71.000000,NaN,NaN


In [11]:
df.columns

Index(['Student_ID', 'Student_Name', 'Age', 'Gender', 'Department', 'Semester',
       'Study_Hours_Per_Day', 'Attendance_Percentage',
       'Previous_Semester_Marks', 'Assignment_Average', 'Internal_Test_Marks',
       'Practical_Marks', 'Participation_Score', 'Sleep_Hours',
       'Internet_Access', 'Study_Resources', 'Backlogs', 'Final_Exam_Marks',
       'Pass_Fail', 'Performance_Category'],
      dtype='object')

Study_Hours_Per_Day
2477 non-null → 63 missing
object        → should probably be numeric

Attendance_Percentage
2465 non-null → 75 missing
object        → should probably be numeric

Step 1 — Check missing values

In [12]:
missing_values = df.isnull().sum()

print(missing_values)

Student_ID                   0
Student_Name                 0
Age                          0
Gender                       0
Department                   0
Semester                     0
Study_Hours_Per_Day         63
Attendance_Percentage       75
Previous_Semester_Marks     50
Assignment_Average          77
Internal_Test_Marks         37
Practical_Marks              0
Participation_Score        110
Sleep_Hours                 75
Internet_Access             26
Study_Resources             76
Backlogs                     0
Final_Exam_Marks            11
Pass_Fail                   11
Performance_Category        11
dtype: int64


In [13]:
missing_percentage = (df.isnull().sum() / len(df)) * 100

missing_summary = pd.DataFrame({
    "Missing_Values": df.isnull().sum(),
    "Missing_Percentage": missing_percentage
})

missing_summary = missing_summary.sort_values(
    by="Missing_Values",
    ascending=False
)

missing_summary

,Missing_Values,Missing_Percentage
Participation_Score,110,4.330709
Assignment_Average,77,3.031496
Study_Resources,76,2.992126
Attendance_Percentage,75,2.952756
Sleep_Hours,75,2.952756
Study_Hours_Per_Day,63,2.480315
Previous_Semester_Marks,50,1.968504
Internal_Test_Marks,37,1.456693
Internet_Access,26,1.023622
Final_Exam_Marks,11,0.433071


Step 2 — Check duplicate rows

In [14]:
duplicate_count = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 34


In [15]:
duplicate_rows = df[df.duplicated()]

duplicate_rows.head()

,Student_ID,Student_Name,Age,Gender,Department,Semester,Study_Hours_Per_Day,Attendance_Percentage,Previous_Semester_Marks,Assignment_Average,Internal_Test_Marks,Practical_Marks,Participation_Score,Sleep_Hours,Internet_Access,Study_Resources,Backlogs,Final_Exam_Marks,Pass_Fail,Performance_Category
7,S23AIM007,SHREYA SAHA,20,Female,CSE (AI & ML),5,2.54,77.2,62.0,13.0,25.5,39,6.0,7.5,Yes,Textbooks,0,62.0,Pass,Good
309,S23CSE107,Tanushree Malhotra,21,Female,Computer Science & Engineering,6,1.5,80.4,42.1,10.5,6.5,31,4.0,7.5,Yes,Online,0,41.0,Pass,Needs Improvement
381,S23CSE178,Sandip Saha,20,Male,Computer Science & Engineering,6,2.59,60.7,74.5,14.5,20.5,34,5.0,5.0,Yes,Online Courses,0,61.0,Pass,Good
385,S23CSE181,Harsh Datta,20,Male,Computer Science & Engineering,6,3.5,90,69.0,17.5,29.0,36,9.0,6.0,Yes,Online Courses,0,76.0,Pass,Very Good
418,S23CSE213,Ayush Malhotra,22,Male,Computer Science & Engineering,6,1.68,83.9,59.0,11.5,17.5,34,6.0,6.0,Yes,Both,1,51.0,Pass,Average


Step 3 — Check unique values of categorical columns

In [16]:
categorical_columns = [
    "Gender",
    "Department",
    "Internet_Access",
    "Study_Resources",
    "Pass_Fail",
    "Performance_Category"
]

for column in categorical_columns:
    print("\n", "=" * 50)
    print(column)
    print(df[column].value_counts(dropna=False))


Gender
Gender
Male       1362
Female      971
Other        33
Male         28
M            27
FEMALE       25
MALE         24
male         23
F            18
Female       16
female       13
Name: count, dtype: int64

Department
Department
Computer Science & Engineering    643
CSE (AI & ML)                     544
Information Technology            415
Electronics & Communication       400
Mechanical Engineering            312
cse                                16
CSE (AIML)                         16
Comp Sci & Engg                    16
Info Tech                          13
E&C                                12
ece                                12
AI & ML                            11
CSE                                11
CSE(AIML)                          11
Computer Science                   10
Mech. Engg                         10
IT                                 10
information technology             10
C.S.E                               9
CSE-AIML                            9


Step 4 — Inspect Study_Hours_Per_Day

Since df.info() showed this as object, let's investigate it.

In [17]:
print(df["Study_Hours_Per_Day"].value_counts(dropna=False).head(30))

Study_Hours_Per_Day
2       411
3       294
2.5     212
1.5     195
4       173
1       164
3.5     109
5        86
4.5      75
NaN      63
6        42
5.5      32
9.5      27
7        16
6.5      16
7.5       7
0.5       7
8         7
2.54      7
2.01      6
3.4       5
1.68      5
1.22      5
2.12      5
1.53      4
1.54      4
1.94      4
3.01      4
1.92      4
2.19      4
Name: count, dtype: int64


In [18]:
print(df["Study_Hours_Per_Day"].unique()[:50])

['3.5' '1' '5.5' '4' '2.94' '2.54' '1.84' '5' '3' '1.5' '3.1' '1.54' '4.5'
 '10' '8.5' '3.73' '2.35' nan '2' '4.84' '1.55' '7.5' '2.65' '3.91' '2.28'
 '2.5' '1.85' '2.98' '1.68' '9.5' '0.84' '1.64' '3.06' '3.11' '1.83' '6'
 '2.97' '3.38' '1.37' '2.43' '1.99' '5.37' '2.6' '3.68' '2.12' '3.59'
 '4.13' '7' '6.27' '3.07']


Step 5 — Inspect Attendance_Percentage

In [19]:
print(df["Attendance_Percentage"].value_counts(dropna=False).head(30))

Attendance_Percentage
NaN     75
100     44
79.3    22
78.6    17
77.2    16
74.3    15
79.8    15
81.6    14
84.7    14
73.9    14
83.9    14
82      14
83      13
73.8    13
80.3    13
79.6    13
77      13
84.9    13
83.4    12
81.7    12
75      12
70.3    12
75.9    12
73.6    12
69.3    12
75.5    11
72.3    11
68.8    11
72.9    11
78.3    11
Name: count, dtype: int64


In [20]:
print(df["Attendance_Percentage"].unique()[:50])

['89.5' '71.5' '78.4' '78.9' '65.4' '75.8' '77.2' '83.2' '74.3' '85.9'
 '79.8' '65.2' '49.7' '68.5' '69.8' '72.6' '80.3' '89' '85.1' '63.5' '82'
 '76.9' '79.5' '69.3' '74.4' '70.1' '91.5' '88.7' '100' '84.9' '98.2'
 '90.4' '79.1' '67.8' '63.9' '90' nan '77.6' '67.7' '48.5' '87.4' '84.4'
 '72.2' '62.7' '57.8' '89.6' '95.2' '79' '73.1' '68.3']


Step 6 — Check numeric columns for suspicious values

In [21]:
numeric_columns = [
    "Age",
    "Semester",
    "Previous_Semester_Marks",
    "Assignment_Average",
    "Internal_Test_Marks",
    "Practical_Marks",
    "Participation_Score",
    "Sleep_Hours",
    "Backlogs",
    "Final_Exam_Marks"
]

df[numeric_columns].describe()

,Age,Semester,Previous_Semester_Marks,Assignment_Average,Internal_Test_Marks,Practical_Marks,Participation_Score,Sleep_Hours,Backlogs,Final_Exam_Marks
count,2540.000000,2540.000000,2490.000000,2463.000000,2503.000000,2540.000000,2430.000000,2465.000000,2540.000000,2529.000000
mean,19.947244,3.869685,63.023655,13.302071,19.543548,33.909055,6.074897,6.709939,0.376772,60.758403
std,3.740548,1.266384,11.516778,4.122158,5.076720,6.402786,1.762139,1.202862,0.743736,19.634813
min,0.000000,2.000000,32.000000,3.000000,4.000000,2.000000,1.000000,0.000000,-1.000000,-7.000000
25%,19.000000,3.000000,55.125000,11.000000,16.500000,30.000000,5.000000,6.000000,0.000000,50.000000
50%,20.000000,4.000000,62.900000,13.000000,19.500000,34.000000,6.000000,6.500000,0.000000,60.000000
75%,21.000000,5.000000,70.800000,15.000000,22.500000,38.000000,7.000000,7.500000,1.000000,71.000000
max,199.000000,6.000000,97.000000,90.000000,78.000000,50.000000,10.000000,25.000000,5.000000,450.000000


In [22]:
df.head(10)

,Student_ID,Student_Name,Age,Gender,Department,Semester,Study_Hours_Per_Day,Attendance_Percentage,Previous_Semester_Marks,Assignment_Average,Internal_Test_Marks,Practical_Marks,Participation_Score,Sleep_Hours,Internet_Access,Study_Resources,Backlogs,Final_Exam_Marks,Pass_Fail,Performance_Category
0,S23AIM001,Tanushree Tiwari,20,Female,CSE (AI & ML),5,3.5,89.5,76.5,19.0,25.0,40,9.0,7.0,Yes,Textbooks,0,77.0,Pass,Very Good
1,S23AIM002,Nitin Kundu,21,Male,CSE (AI & ML),5,1,71.5,32.0,8.0,4.0,22,7.0,7.0,Yes,Online Courses,4,16.0,Fail,Needs Improvement
2,S23AIM003,Amit Laha,21,Male,CSE (AI & ML),5,5.5,78.4,82.0,20.0,26.5,45,5.0,6.0,Yes,Notes Only,2,80.0,Pass,Very Good
3,S23AIM004,Akash Thakur,21,Male,CSE (AI & ML),6,4,78.9,67.7,16.5,21.5,38,6.0,5.0,Yes,Online Courses,0,63.0,Pass,Good
4,S23AIM005,Aarav Rao,22,Male,cse aiml,6,2.94,65.4,49.0,10.0,13.0,34,6.0,5.0,Yes,Both,0,47.0,Pass,Needs Improvement
5,S23AIM006,Deepak Kundu,22,Male,CSE (AI & ML),6,4,75.8,64.0,20.0,29.0,35,9.0,5.5,No,Both,0,87.0,Pass,Very Good
6,S23AIM007,SHREYA SAHA,20,Female,CSE (AI & ML),5,2.54,77.2,62.0,13.0,25.5,39,6.0,7.5,Yes,Textbooks,0,62.0,Pass,Good
7,S23AIM007,SHREYA SAHA,20,Female,CSE (AI & ML),5,2.54,77.2,62.0,13.0,25.5,39,6.0,7.5,Yes,Textbooks,0,62.0,Pass,Good
8,S23AIM008,Vaishnavi Ansari,22,Female,CSE (AI & ML),5,1.84,83.2,71.0,15.0,23.0,32,7.0,6.5,Yes,Both,1,70.0,Pass,Good
9,S23AIM009,Suraj Das,22,Male,CSE (AI & ML),5,3.5,74.3,66.5,15.5,16.5,33,7.0,6.0,Yes,Both,0,69.0,Pass,Good


Check suspicious ages

In [23]:
# Check suspicious Age values

print(df[df["Age"] <= 0][["Student_ID", "Student_Name", "Age"]])

print("\nAges greater than 100:")
print(df[df["Age"] > 100][["Student_ID", "Student_Name", "Age"]])

     Student_ID  Student_Name  Age
2085   S24ME087  Kartik Kumar    0

Ages greater than 100:
     Student_ID Student_Name  Age
1715  S24ECE162  Rahul Yadav  199


Check suspicious final exam marks

In [24]:
# Check suspicious Final_Exam_Marks values

print(df[df["Final_Exam_Marks"] < 0][
    ["Student_ID", "Student_Name", "Final_Exam_Marks"]
])

print("\nFinal Exam Marks greater than 100:")
print(df[df["Final_Exam_Marks"] > 100][
    ["Student_ID", "Student_Name", "Final_Exam_Marks"]
])

     Student_ID     Student_Name  Final_Exam_Marks
23    S23AIM023         Ari Khan              -7.0
1101  S24AIM227  Rituparna Gupta              -7.0
2186  S25AIM025       Tanya Paul              -7.0

Final Exam Marks greater than 100:
     Student_ID         Student_Name  Final_Exam_Marks
32    S23AIM032           Sakshi Rao             450.0
441   S23ECE002           Tanmay Pal             135.0
804    S23ME045        RITWIK ANSARI             110.0
1163  S24AIM289       Rituparna Bose             450.0
1784   S24IT013      Md Faisal Reddy             135.0
2048   S24ME050  Aarushi Purkayastha             135.0


Check suspicious sleep hours

In [25]:
# Check suspicious Sleep_Hours values

print(df[df["Sleep_Hours"] < 0][
    ["Student_ID", "Student_Name", "Sleep_Hours"]
])

print("\nSleep hours greater than 24:")
print(df[df["Sleep_Hours"] > 24][
    ["Student_ID", "Student_Name", "Sleep_Hours"]
])

Empty DataFrame
Columns: [Student_ID, Student_Name, Sleep_Hours]
Index: []

Sleep hours greater than 24:
     Student_ID Student_Name  Sleep_Hours
1014  S24AIM140    Sumit Pal         25.0


Check all numeric ranges in one place

In [26]:
for column in numeric_columns:
    print("\n", column)
    print(df[column].sort_values().head(5).to_list())
    print("Highest values:", df[column].sort_values(ascending=False).head(5).to_list())


 Age
[0, 2, 2, 18, 18]
Highest values: [199, 22, 22, 22, 22]

 Semester
[2, 2, 2, 2, 2]
Highest values: [6, 6, 6, 6, 6]

 Previous_Semester_Marks
[32.0, 32.0, 32.0, 32.0, 32.0]
Highest values: [97.0, 97.0, 97.0, 97.0, 97.0]

 Assignment_Average
[3.0, 4.0, 5.0, 5.0, 5.5]
Highest values: [90.0, 85.0, 85.0, 85.0, 20.0]

 Internal_Test_Marks
[4.0, 4.0, 4.0, 6.0, 6.5]
Highest values: [78.0, 78.0, 78.0, 30.0, 30.0]

 Practical_Marks
[2, 2, 5, 5, 12]
Highest values: [50, 50, 50, 50, 50]

 Participation_Score
[1.0, 1.0, 1.0, 1.0, 1.0]
Highest values: [10.0, 10.0, 10.0, 10.0, 10.0]

 Sleep_Hours
[0.0, 0.0, 3.5, 3.5, 3.5]
Highest values: [25.0, 20.0, 20.0, 10.0, 10.0]

 Backlogs
[-1, -1, -1, 0, 0]
Highest values: [5, 5, 5, 5, 5]

 Final_Exam_Marks
[-7.0, -7.0, -7.0, 8.0, 8.0]
Highest values: [450.0, 450.0, 135.0, 135.0, 135.0]


In [27]:
df["Study_Hours_Per_Day"].unique()

array(['3.5', '1', '5.5', '4', '2.94', '2.54', '1.84', '5', '3', '1.5',
       '3.1', '1.54', '4.5', '10', '8.5', '3.73', '2.35', nan, '2',
       '4.84', '1.55', '7.5', '2.65', '3.91', '2.28', '2.5', '1.85',
       '2.98', '1.68', '9.5', '0.84', '1.64', '3.06', '3.11', '1.83', '6',
       '2.97', '3.38', '1.37', '2.43', '1.99', '5.37', '2.6', '3.68',
       '2.12', '3.59', '4.13', '7', '6.27', '3.07', '0.5', '2.05', '2.81',
       '1.8', '4.3', '1.22', '2.55', '1.61', '5.38', '9', '1.19', '1.01',
       '2.2', '1.52', '2.8', '3.72', '5.47', '6.26', '2.83', '4.44',
       '4.19', '1.2', '3.23', '2.34', '1.25', '1.18', '2.58', '2.75',
       '3.64', '4.03', '2.25', '2.64', '3.26', '5.24', '3.62', '4.71',
       '1.86', '5.02', '2.13', '2.68', '8.9', '3.85', '2.11', '1.94',
       '2.5 hrs', '2.57', '8.83', '3.6', '3.45', '0.78', '2.23', '5.42',
       '1.62', '4.14', '0.94', '3.4', '1.56', '2.59', '3.71', '6.17', '8',
       '3.92', '5.81', '6.5', '4.31', '1.36', '3.01', '4.74', '0.68',

In [28]:
df["Attendance_Percentage"].unique()

array(['89.5', '71.5', '78.4', '78.9', '65.4', '75.8', '77.2', '83.2',
       '74.3', '85.9', '79.8', '65.2', '49.7', '68.5', '69.8', '72.6',
       '80.3', '89', '85.1', '63.5', '82', '76.9', '79.5', '69.3', '74.4',
       '70.1', '91.5', '88.7', '100', '84.9', '98.2', '90.4', '79.1',
       '67.8', '63.9', '90', nan, '77.6', '67.7', '48.5', '87.4', '84.4',
       '72.2', '62.7', '57.8', '89.6', '95.2', '79', '73.1', '68.3',
       '70.5', '81', '71.8', '83.9', '55.1', '74.2', '62.2', '87.3',
       '73.6', '75.9', '83', '57.7', '85.6', '80.5', '75.5', '83.7',
       '75.7', '83.6', '84', '82.3', '56.4', '68.6', '82.5', '85.5',
       '44.7', '69.4', '64.3', '74.8', '59.9', '79.3', '84.3', '57.2',
       '66.4', '77.5', '80.9', '84.5', '66', '75', '82.9', '81.8', '56.8',
       '60.5', '73.2', '86.4', '66.9', '73.7', '69.1', '82.8', '95.6',
       '85.8', '81.1', '78.3', '74.5', '67.3', '60.8', '86.5', '78.7',
       '67.1', '83.4', '76.8', '69.6', '75.3', '60.1', '52.1', '68.1',
    

In [29]:
print("Unique values:")
print(df["Attendance_Percentage"].unique())

print("\nValue counts:")
print(df["Attendance_Percentage"].value_counts(dropna=False))

Unique values:
['89.5' '71.5' '78.4' '78.9' '65.4' '75.8' '77.2' '83.2' '74.3' '85.9'
 '79.8' '65.2' '49.7' '68.5' '69.8' '72.6' '80.3' '89' '85.1' '63.5' '82'
 '76.9' '79.5' '69.3' '74.4' '70.1' '91.5' '88.7' '100' '84.9' '98.2'
 '90.4' '79.1' '67.8' '63.9' '90' nan '77.6' '67.7' '48.5' '87.4' '84.4'
 '72.2' '62.7' '57.8' '89.6' '95.2' '79' '73.1' '68.3' '70.5' '81' '71.8'
 '83.9' '55.1' '74.2' '62.2' '87.3' '73.6' '75.9' '83' '57.7' '85.6'
 '80.5' '75.5' '83.7' '75.7' '83.6' '84' '82.3' '56.4' '68.6' '82.5'
 '85.5' '44.7' '69.4' '64.3' '74.8' '59.9' '79.3' '84.3' '57.2' '66.4'
 '77.5' '80.9' '84.5' '66' '75' '82.9' '81.8' '56.8' '60.5' '73.2' '86.4'
 '66.9' '73.7' '69.1' '82.8' '95.6' '85.8' '81.1' '78.3' '74.5' '67.3'
 '60.8' '86.5' '78.7' '67.1' '83.4' '76.8' '69.6' '75.3' '60.1' '52.1'
 '68.1' '61.1' '73.5' '73.3' '76.4' '66.6' '83.5' '81.7' '58.4' '88.8'
 '64' '80.1' '91.4' '69.5' '64.2' '82.7' '80.2' '77' '91.8' '73.9' '77.7'
 '77.9' '78.1' '70.4' '89.8' '94.1' '75.1' '85.2' '71

## Finding: Attendance_Percentage

The `Attendance_Percentage` column was detected as an object data type instead of numeric.

During inspection, the following data-quality issues were found:

- Numeric percentages are stored as strings.
- Some values contain the `%` symbol.
- Missing values are present.
- Negative attendance values are present.
- Some values are greater than 100%.

The column therefore needs to be cleaned and converted to a numeric data type.
Values outside the valid attendance range of 0–100 will be treated as invalid.

Check duplicate rows

In [30]:
# Check duplicate rows

duplicate_rows = df.duplicated().sum()

print("Duplicate rows:", duplicate_rows)

duplicate_student_ids = df["Student_ID"].duplicated().sum()

print("Duplicate Student IDs:", duplicate_student_ids)

Duplicate rows: 34
Duplicate Student IDs: 40


## Finding: Duplicate Records

The dataset contains:

- 34 duplicate rows
- 40 duplicate Student IDs

Duplicate rows represent records that are completely repeated.

Duplicate Student IDs require further investigation because two records with the same Student_ID may contain different information. Therefore, duplicate Student IDs will not be removed automatically without checking the corresponding records.

Investigating the duplicate Student IDs

In [31]:
# Find Student IDs that occur more than once

duplicate_ids = df[
    df["Student_ID"].duplicated(keep=False)
]["Student_ID"].value_counts()

print(duplicate_ids)

Student_ID
S23AIM007    2
S23CSE107    2
S23CSE178    2
S23CSE181    2
S23CSE213    2
S23CSE218    2
S23CSE231    2
S23CSE232    2
S23ECE020    2
S23ECE055    2
S23ECE080    2
S23ECE097    2
S23ECE113    2
S23ME038     2
S24AIM150    2
S24AIM188    2
S24AIM305    2
S24CSE013    2
S24CSE145    2
S24CSE298    2
S24CSE349    2
S24ECE124    2
S24ECE180    2
S24ECE184    2
S24IT118     2
S24IT138     2
S24IT158     2
S24IT164     2
S24ME004     2
S24ME011     2
S24ME054     2
S24ME108     2
S25AIM035    2
S25CSE032    2
S25ECE001    2
S25ECE016    2
S25ECE020    2
S25ME012     2
S25ME013     2
S25ME025     2
Name: count, dtype: int64


In [32]:
# Display records having duplicate Student IDs

duplicate_id_list = duplicate_ids.index

duplicate_id_records = df[
    df["Student_ID"].isin(duplicate_id_list)
].sort_values("Student_ID")

duplicate_id_records

,Student_ID,Student_Name,Age,Gender,Department,Semester,Study_Hours_Per_Day,Attendance_Percentage,Previous_Semester_Marks,Assignment_Average,Internal_Test_Marks,Practical_Marks,Participation_Score,Sleep_Hours,Internet_Access,Study_Resources,Backlogs,Final_Exam_Marks,Pass_Fail,Performance_Category
6,S23AIM007,SHREYA SAHA,20,Female,CSE (AI & ML),5,2.54,77.2,62.0,13.0,25.5,39,6.0,7.5,Yes,Textbooks,0,62.0,Pass,Good
7,S23AIM007,SHREYA SAHA,20,Female,CSE (AI & ML),5,2.54,77.2,62.0,13.0,25.5,39,6.0,7.5,Yes,Textbooks,0,62.0,Pass,Good
308,S23CSE107,Tanushree Malhotra,21,Female,Computer Science & Engineering,6,1.5,80.4,42.1,10.5,6.5,31,4.0,7.5,Yes,Online,0,41.0,Pass,Needs Improvement
309,S23CSE107,Tanushree Malhotra,21,Female,Computer Science & Engineering,6,1.5,80.4,42.1,10.5,6.5,31,4.0,7.5,Yes,Online,0,41.0,Pass,Needs Improvement
380,S23CSE178,Sandip Saha,20,Male,Computer Science & Engineering,6,2.59,60.7,74.5,14.5,20.5,34,5.0,5.0,Yes,Online Courses,0,61.0,Pass,Good
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2471,S25ME012,Aarav Biswas,19,Male,Mechanical Engineering,2,9.5,99.8,74.0,17.5,24.0,37,9.0,8.5,No,NaN,0,81.0,Pass,Very Good
2474,S25ME013,Poulami Bhattacharya,18,Female,ME,2,3,65,89.7,13.5,21.0,29,7.0,8.0,Yes,Both,0,64.0,Pass,Good
2473,S25ME013,Poulami Bhattacharya,18,Female,ME,2,3,65,89.7,13.5,21.0,29,7.0,8.0,Yes,Both,0,64.0,Pass,Good
2522,S25ME025,Isha Datta,19,Female,Mechanical Engineering,2,3.06,76.4,70.0,13.5,18.0,33,NaN,7.0,Yes,Online Courses,0,73.0,Pass,Good


In [33]:
# Check whether duplicate Student IDs have conflicting records

duplicate_id_records = df[
    df["Student_ID"].duplicated(keep=False)
].sort_values("Student_ID")

duplicate_id_summary = duplicate_id_records.groupby("Student_ID").size()

print("Number of unique duplicated Student IDs:", len(duplicate_id_summary))
print("\nDuplicate count distribution:")
print(duplicate_id_summary.value_counts())

Number of unique duplicated Student IDs: 40

Duplicate count distribution:
2    40
Name: count, dtype: int64


In [34]:
# Check duplicate Student IDs against complete-row duplicates

duplicate_id_df = df[df["Student_ID"].duplicated(keep=False)]

conflicting_ids = []

for student_id, group in duplicate_id_df.groupby("Student_ID"):
    if len(group.drop_duplicates()) > 1:
        conflicting_ids.append(student_id)

print("Student IDs with conflicting records:", len(conflicting_ids))
print(conflicting_ids)

Student IDs with conflicting records: 6
['S24AIM150', 'S24AIM188', 'S24CSE145', 'S24CSE349', 'S24ECE124', 'S24ECE184']


## Finding: Duplicate Student IDs

The dataset contains 40 Student IDs that occur twice.

Further investigation showed:

- 34 duplicated Student IDs correspond to completely identical records.
- 6 duplicated Student IDs contain conflicting records.
- Therefore, duplicate Student IDs cannot all be removed automatically.

The 34 exact duplicate records can be safely removed.

The 6 conflicting Student IDs require separate investigation before deciding which records to retain.

In [35]:
conflicting_records = df[
    df["Student_ID"].isin(conflicting_ids)
].sort_values("Student_ID")

conflicting_records

,Student_ID,Student_Name,Age,Gender,Department,Semester,Study_Hours_Per_Day,Attendance_Percentage,Previous_Semester_Marks,Assignment_Average,Internal_Test_Marks,Practical_Marks,Participation_Score,Sleep_Hours,Internet_Access,Study_Resources,Backlogs,Final_Exam_Marks,Pass_Fail,Performance_Category
1024,S24AIM150,Aarav Kundu,19,Male,CSE (AI & ML),3,4.51,75.7,66.4,NaN,20.5,32,7.0,6.0,Yes,Online Courses,0,76.0,Pass,Very Good
2538,S24AIM150,AARAV KUNDU,19,Male,CSE (AI & ML),3,4.51,75.7,66.4,NaN,20.5,32,7.0,6.0,Yes,Online Courses,0,76.0,Pass,Very Good
1062,S24AIM188,Moumita Basu,19,Female,CSE (AI & ML),3,1,71,46.8,8.0,13.0,36,7.0,8.5,Yes,Both,2,28.0,Fail,Needs Improvement
2535,S24AIM188,MOUMITA BASU,19,Female,CSE (AI & ML),3,1,71,46.8,8.0,13.0,36,7.0,8.5,Yes,Both,2,28.0,Fail,Needs Improvement
1333,S24CSE145,Harsh Ansari,19,Male,Computer Science & Engineering,4,NaN,92.9,75.0,14.5,23.0,41,7.0,8.5,Yes,Both,0,85.0,Pass,Very Good
2536,S24CSE145,HARSH ANSARI,19,Male,Computer Science & Engineering,4,NaN,92.9,75.0,14.5,23.0,41,7.0,8.5,Yes,Both,0,85.0,Pass,Very Good
1538,S24CSE349,Rishabh Banerjee,20,Male,Computer Science & Engineering,4,1.5,69.2,54.0,12.5,15.0,24,2.0,7.0,Yes,Online Courses,3,32.0,Fail,Needs Improvement
2537,S24CSE349,RISHABH BANERJEE,20,Male,Computer Science & Engineering,4,1.5,69.2,54.0,12.5,15.0,24,2.0,7.0,Yes,Online Courses,3,32.0,Fail,Needs Improvement
1677,S24ECE124,Rishabh Chakraborty,19,Male,Electronics & Comm.,3,2.5,84.7,69.0,13.5,22.0,35,8.0,6.5,NaN,Both,1,57.0,Pass,Average
2534,S24ECE124,RISHABH CHAKRABORTY,19,Male,Electronics & Comm.,3,2.5,84.7,69.0,13.5,22.0,35,8.0,6.5,NaN,Both,1,57.0,Pass,Average


In [36]:
for student_id in conflicting_ids:
    print("\n" + "=" * 80)
    print("Student ID:", student_id)
    print("=" * 80)
    
    print(
        df[df["Student_ID"] == student_id].to_string(index=False)
    )


Student ID: S24AIM150
Student_ID Student_Name  Age Gender    Department  Semester Study_Hours_Per_Day Attendance_Percentage  Previous_Semester_Marks  Assignment_Average  Internal_Test_Marks  Practical_Marks  Participation_Score  Sleep_Hours Internet_Access Study_Resources  Backlogs  Final_Exam_Marks Pass_Fail Performance_Category
 S24AIM150  Aarav Kundu   19   Male CSE (AI & ML)         3                4.51                  75.7                     66.4                 NaN                 20.5               32                  7.0          6.0             Yes  Online Courses         0              76.0      Pass            Very Good
 S24AIM150  AARAV KUNDU   19   Male CSE (AI & ML)         3                4.51                  75.7                     66.4                 NaN                 20.5               32                  7.0          6.0             Yes  Online Courses         0              76.0      Pass            Very Good

Student ID: S24AIM188
Student_ID Student_Name 

Student_ID Student_Name  Age Gender                     Department  Semester Study_Hours_Per_Day Attendance_Percentage  Previous_Semester_Marks  Assignment_Average  Internal_Test_Marks  Practical_Marks  Participation_Score  Sleep_Hours Internet_Access Study_Resources  Backlogs  Final_Exam_Marks Pass_Fail Performance_Category
 S24CSE145 Harsh Ansari   19   Male Computer Science & Engineering         4                 NaN                  92.9                     75.0                14.5                 23.0               41                  7.0          8.5             Yes            Both         0              85.0      Pass            Very Good
 S24CSE145 HARSH ANSARI   19   Male Computer Science & Engineering         4                 NaN                  92.9                     75.0                14.5                 23.0               41                  7.0          8.5             Yes            Both         0              85.0      Pass            Very Good

Student ID: S24CSE

## Finding: Duplicate Records and Student IDs

The dataset initially contained 34 exact duplicate rows and 40 duplicated Student IDs.

Further investigation showed that all 40 duplicated Student IDs represented duplicate student records.

For 34 Student IDs, the duplicate records were completely identical.

For the remaining 6 Student IDs, the only difference was the capitalization of the student's name. For example, "Aarav Kundu" and "AARAV KUNDU" represented the same student and had identical values in all other columns.

Therefore, these records were identified as redundant duplicates and will be removed during the cleaning process.

In [37]:
# Create a copy of the original dataframe for cleaning

df_cleaned = df.copy()

print("Original number of rows:", len(df_cleaned))

Original number of rows: 2540


Step 1 — Standardize student names

In [38]:
# Standardize student names

df_cleaned["Student_Name"] = (
    df_cleaned["Student_Name"]
    .str.strip()
    .str.title()
)

print(df_cleaned["Student_Name"].head())

0    Tanushree Tiwari
1         Nitin Kundu
2           Amit Laha
3        Akash Thakur
4           Aarav Rao
Name: Student_Name, dtype: object


Step 2 — Remove duplicate rows

In [39]:
# Remove duplicate records

before_duplicates = len(df_cleaned)

df_cleaned = df_cleaned.drop_duplicates()

after_duplicates = len(df_cleaned)

print("Rows before removing duplicates:", before_duplicates)
print("Rows after removing duplicates:", after_duplicates)
print("Duplicate rows removed:", before_duplicates - after_duplicates)

Rows before removing duplicates: 2540
Rows after removing duplicates: 2500
Duplicate rows removed: 40


In [40]:
print("Remaining duplicate rows:", df_cleaned.duplicated().sum())
print("Remaining duplicate Student IDs:",
      df_cleaned["Student_ID"].duplicated().sum())

Remaining duplicate rows: 0
Remaining duplicate Student IDs: 0


Step 1 — Make a copy of the column

In [41]:
study_hours = df_cleaned["Study_Hours_Per_Day"].copy()

Step 2 — Remove "hrs" and convert to numeric

In [42]:
df_cleaned["Study_Hours_Per_Day"] = (
    df_cleaned["Study_Hours_Per_Day"]
    .astype("string")
    .str.replace("hrs", "", case=False, regex=False)
    .str.strip()
)

df_cleaned["Study_Hours_Per_Day"] = pd.to_numeric(
    df_cleaned["Study_Hours_Per_Day"],
    errors="coerce"
)

In [43]:
df_cleaned["Study_Hours_Per_Day"] = (
    df_cleaned["Study_Hours_Per_Day"]
    .astype("string")
    .str.replace("hrs", "", case=False, regex=False)
    .str.strip()
)

df_cleaned["Study_Hours_Per_Day"] = pd.to_numeric(
    df_cleaned["Study_Hours_Per_Day"],
    errors="coerce"
)

In [44]:
print(df_cleaned["Study_Hours_Per_Day"].dtype)
print(df_cleaned["Study_Hours_Per_Day"].describe())

Float64
count      2438.0
mean     3.000968
std      1.726081
min          -2.0
25%           2.0
50%           2.5
75%          3.95
max          18.0
Name: Study_Hours_Per_Day, dtype: Float64


Step 3 — Find invalid study-hour values

In [45]:
invalid_study_hours = df_cleaned[
    (df_cleaned["Study_Hours_Per_Day"] < 0) |
    (df_cleaned["Study_Hours_Per_Day"] > 24)
]

print("Invalid study-hour records:", len(invalid_study_hours))

invalid_study_hours[
    ["Student_ID", "Study_Hours_Per_Day"]
]

Invalid study-hour records: 3


,Student_ID,Study_Hours_Per_Day
907,S24AIM033,-2.0
1902,S24IT131,-1.5
2112,S24ME114,-1.5


Step 4 — Replace invalid study hours with NaN

In [46]:
# Replace invalid study-hour values with NaN

invalid_mask = (
    (df_cleaned["Study_Hours_Per_Day"] < 0) |
    (df_cleaned["Study_Hours_Per_Day"] > 24)
)

df_cleaned.loc[invalid_mask, "Study_Hours_Per_Day"] = np.nan

print("Invalid values replaced with NaN:",
      invalid_mask.sum())

Invalid values replaced with NaN: 3


Step 5 — Check missing values

In [47]:
print(
    "Missing Study_Hours_Per_Day:",
    df_cleaned["Study_Hours_Per_Day"].isna().sum()
)

Missing Study_Hours_Per_Day: 65


In [48]:
df_cleaned["Study_Hours_Per_Day"].describe()

count      2435.0
mean     3.006719
std      1.719323
min           0.0
25%           2.0
50%           2.5
75%         3.965
max          18.0
Name: Study_Hours_Per_Day, dtype: Float64

## Cleaning: Study_Hours_Per_Day

The `Study_Hours_Per_Day` column was originally stored as an object because some values contained the text `"hrs"`.

The following cleaning steps were performed:

1. Removed the `"hrs"` text from values.
2. Converted the column to a numeric data type.
3. Identified 3 negative values as invalid.
4. Replaced the 3 invalid negative values with `NaN`.
5. Kept missing values as `NaN` for later missing-value treatment.

After cleaning, the column contains 2,435 valid numeric values and 65 missing values.

Clean Attendance_Percentage

Step 1 — Remove % and convert to numeric

In [49]:
# Clean Attendance_Percentage

df_cleaned["Attendance_Percentage"] = (
    df_cleaned["Attendance_Percentage"]
    .astype("string")
    .str.replace("%", "", regex=False)
    .str.strip()
)

df_cleaned["Attendance_Percentage"] = pd.to_numeric(
    df_cleaned["Attendance_Percentage"],
    errors="coerce"
)

print(df_cleaned["Attendance_Percentage"].dtype)

Float64


Step 2 — Find invalid attendance values

In [50]:
# Find invalid attendance values

invalid_attendance = df_cleaned[
    (df_cleaned["Attendance_Percentage"] < 0) |
    (df_cleaned["Attendance_Percentage"] > 100)
]

print("Invalid attendance records:", len(invalid_attendance))

invalid_attendance[
    ["Student_ID", "Attendance_Percentage"]
]

Invalid attendance records: 13


,Student_ID,Attendance_Percentage
469,S23ECE029,101.0
625,S23IT036,-12.0
831,S23ME072,112.0
935,S24AIM061,150.0
1130,S24AIM256,-5.0
1133,S24AIM259,112.0
1258,S24CSE070,120.0
1432,S24CSE244,-5.0
1570,S24ECE017,-5.0
1709,S24ECE156,101.0


Step 3 — Replace invalid attendance values with NaN

In [51]:
# Replace invalid attendance values with NaN

invalid_attendance_mask = (
    (df_cleaned["Attendance_Percentage"] < 0) |
    (df_cleaned["Attendance_Percentage"] > 100)
)

df_cleaned.loc[
    invalid_attendance_mask,
    "Attendance_Percentage"
] = np.nan

print(
    "Invalid attendance values replaced with NaN:",
    invalid_attendance_mask.sum()
)

Invalid attendance values replaced with NaN: 13


Step 4 — Check missing attendance values

In [52]:
print(
    "Missing Attendance_Percentage:",
    df_cleaned["Attendance_Percentage"].isna().sum()
)

Missing Attendance_Percentage: 88


In [53]:
df_cleaned["Attendance_Percentage"].describe()

count       2412.0
mean     77.072388
std      10.832225
min           38.0
25%           69.5
50%           77.0
75%           84.6
max          100.0
Name: Attendance_Percentage, dtype: Float64

## Cleaning Result: Attendance_Percentage

After cleaning, the `Attendance_Percentage` column contains valid numeric values between 0 and 100.

The 13 invalid values that were below 0 or above 100 were replaced with `NaN`.

The cleaned column has:
- 2,412 valid values
- 88 missing values
- Minimum: 38.0
- Maximum: 100.0
- Mean: approximately 77.07

# Check Age

In [54]:
# Check suspicious Age values

invalid_age = df_cleaned[
    (df_cleaned["Age"] < 15) |
    (df_cleaned["Age"] > 30)
]

print("Suspicious age records:", len(invalid_age))

invalid_age[
    ["Student_ID", "Student_Name", "Age"]
]

Suspicious age records: 4


,Student_ID,Student_Name,Age
1715,S24ECE162,Rahul Yadav,199
2085,S24ME087,Kartik Kumar,0
2199,S25AIM037,Aditi Datta,2
2293,S25CSE051,Sakshi Roy,2


Step 1 — Replace invalid ages

In [55]:
# Replace invalid age values with NaN

invalid_age_mask = (
    (df_cleaned["Age"] < 15) |
    (df_cleaned["Age"] > 30)
)

df_cleaned.loc[
    invalid_age_mask,
    "Age"
] = np.nan

print(
    "Invalid age values replaced with NaN:",
    invalid_age_mask.sum()
)

Invalid age values replaced with NaN: 4


Step 2 — Verify

In [56]:
print("Missing Age:", df_cleaned["Age"].isna().sum())

print("\nAge statistics:")
print(df_cleaned["Age"].describe())

Missing Age: 4

Age statistics:
count    2496.000000
mean       19.899439
std         0.976633
min        18.000000
25%        19.000000
50%        20.000000
75%        21.000000
max        22.000000
Name: Age, dtype: float64


## Cleaning Result: Age

Four invalid age values were identified and replaced with `NaN`.

After cleaning:

- Valid age range: 18–22
- Valid records: 2,496
- Missing values: 4
- Mean age: approximately 19.90

##  Check Sleep_Hours

In [57]:
# Check Sleep_Hours values

print("Unique Sleep_Hours values:")
print(sorted(df_cleaned["Sleep_Hours"].dropna().unique()))

print("\nMissing Sleep_Hours:",
      df_cleaned["Sleep_Hours"].isna().sum())

Unique Sleep_Hours values:
[np.float64(0.0), np.float64(3.5), np.float64(4.0), np.float64(4.5), np.float64(5.0), np.float64(5.5), np.float64(6.0), np.float64(6.5), np.float64(7.0), np.float64(7.5), np.float64(8.0), np.float64(8.5), np.float64(9.0), np.float64(9.5), np.float64(10.0), np.float64(20.0), np.float64(25.0)]

Missing Sleep_Hours: 75


In [58]:
# Check for invalid Sleep_Hours

invalid_sleep = df_cleaned[
    (df_cleaned["Sleep_Hours"] < 0) |
    (df_cleaned["Sleep_Hours"] > 24)
]

print("Invalid Sleep_Hours records:", len(invalid_sleep))

invalid_sleep[
    ["Student_ID", "Student_Name", "Sleep_Hours"]
]

Invalid Sleep_Hours records: 1


,Student_ID,Student_Name,Sleep_Hours
1014,S24AIM140,Sumit Pal,25.0


In [59]:
# Inspect unusually high sleep-hour values

unusual_sleep = df_cleaned[
    df_cleaned["Sleep_Hours"] >= 20
]

unusual_sleep[
    ["Student_ID", "Student_Name", "Sleep_Hours"]
]

,Student_ID,Student_Name,Sleep_Hours
1014,S24AIM140,Sumit Pal,25.0
2001,S24ME004,Rajesh Sharma,20.0


In [60]:
# Inspect zero sleep-hour values

zero_sleep = df_cleaned[
    df_cleaned["Sleep_Hours"] == 0
]

print("Zero sleep-hour records:", len(zero_sleep))

zero_sleep[
    ["Student_ID", "Student_Name", "Sleep_Hours"]
]

Zero sleep-hour records: 2


,Student_ID,Student_Name,Sleep_Hours
962,S24AIM088,Moumita Ansari,0.0
2084,S24ME086,Kartik Mondal,0.0


Step 1 — Replace invalid sleep values

In [61]:
# Replace invalid/suspicious sleep-hour values with NaN

invalid_sleep_mask = (
    (df_cleaned["Sleep_Hours"] <= 0) |
    (df_cleaned["Sleep_Hours"] > 12)
)

df_cleaned.loc[
    invalid_sleep_mask,
    "Sleep_Hours"
] = np.nan

print(
    "Invalid/suspicious sleep values replaced with NaN:",
    invalid_sleep_mask.sum()
)

Invalid/suspicious sleep values replaced with NaN: 4


Step 2 — Check the result

In [62]:
print("Missing Sleep_Hours:",
      df_cleaned["Sleep_Hours"].isna().sum())

print("\nSleep_Hours statistics:")
print(df_cleaned["Sleep_Hours"].describe())

Missing Sleep_Hours: 79

Sleep_Hours statistics:
count    2421.000000
mean        6.697233
std         1.062815
min         3.500000
25%         6.000000
50%         6.500000
75%         7.500000
max        10.000000
Name: Sleep_Hours, dtype: float64


## Cleaning Result: Sleep_Hours

The `Sleep_Hours` column originally contained 75 missing values.

During validation, 4 suspicious values were identified:
- Two records with 0 hours
- One record with 20 hours
- One record with 25 hours

These values were treated as invalid/suspicious and replaced with `NaN`.

After cleaning:
- Valid values: 2,421
- Missing values: 79
- Minimum: 3.5 hours
- Maximum: 10.0 hours
- Mean: approximately 6.70 hours

## Final_Exam_Marks

In [63]:
# Check invalid Final Exam Marks

invalid_final_marks = df_cleaned[
    (df_cleaned["Final_Exam_Marks"] < 0) |
    (df_cleaned["Final_Exam_Marks"] > 100)
]

print("Invalid Final_Exam_Marks records:", len(invalid_final_marks))

invalid_final_marks[
    ["Student_ID", "Student_Name", "Final_Exam_Marks", "Pass_Fail", "Performance_Category"]
]

Invalid Final_Exam_Marks records: 9


,Student_ID,Student_Name,Final_Exam_Marks,Pass_Fail,Performance_Category
23,S23AIM023,Ari Khan,-7.0,Pass,Needs Improvement
32,S23AIM032,Sakshi Rao,450.0,Pass,Average
441,S23ECE002,Tanmay Pal,135.0,Pass,Excellent
804,S23ME045,Ritwik Ansari,110.0,Pass,Needs Improvement
1101,S24AIM227,Rituparna Gupta,-7.0,Pass,Good
1163,S24AIM289,Rituparna Bose,450.0,Pass,Very Good
1784,S24IT013,Md Faisal Reddy,135.0,Pass,Good
2048,S24ME050,Aarushi Purkayastha,135.0,Pass,Average
2186,S25AIM025,Tanya Paul,-7.0,Pass,Good


Step 1 — Replace invalid marks with NaN

In [64]:
# Replace invalid Final_Exam_Marks with NaN

invalid_final_marks_mask = (
    (df_cleaned["Final_Exam_Marks"] < 0) |
    (df_cleaned["Final_Exam_Marks"] > 100)
)

df_cleaned.loc[
    invalid_final_marks_mask,
    "Final_Exam_Marks"
] = np.nan

print(
    "Invalid Final_Exam_Marks replaced with NaN:",
    invalid_final_marks_mask.sum()
)

Invalid Final_Exam_Marks replaced with NaN: 9


Step 2 — Verify

In [65]:
print(
    "Missing Final_Exam_Marks:",
    df_cleaned["Final_Exam_Marks"].isna().sum()
)

print("\nFinal_Exam_Marks statistics:")
print(df_cleaned["Final_Exam_Marks"].describe())

Missing Final_Exam_Marks: 20

Final_Exam_Marks statistics:
count    2480.000000
mean       60.450403
std        15.925374
min         8.000000
25%        50.000000
50%        60.000000
75%        71.000000
max       100.000000
Name: Final_Exam_Marks, dtype: float64


## Cleaning Result: Final_Exam_Marks

The `Final_Exam_Marks` column contained 9 invalid values outside the valid 0–100 range.

These invalid values were replaced with `NaN`.

After cleaning:
- Valid values: 2,480
- Missing values: 20
- Minimum: 8
- Maximum: 100
- Mean: approximately 60.45

## Previous_Semester_Marks

In [66]:
# Check Previous_Semester_Marks

print("Missing values:",
      df_cleaned["Previous_Semester_Marks"].isna().sum())

invalid_previous_marks = df_cleaned[
    (df_cleaned["Previous_Semester_Marks"] < 0) |
    (df_cleaned["Previous_Semester_Marks"] > 100)
]

print(
    "Invalid Previous_Semester_Marks records:",
    len(invalid_previous_marks)
)

invalid_previous_marks[
    ["Student_ID", "Student_Name", "Previous_Semester_Marks"]
]

Missing values: 50
Invalid Previous_Semester_Marks records: 0


,Student_ID,Student_Name,Previous_Semester_Marks


## Cleaning Result: Previous_Semester_Marks

The `Previous_Semester_Marks` column was checked for values outside the valid range of 0–100.

No invalid values were found.

The column contains:
- 50 missing values
- 0 invalid values
- All non-missing values within the valid range

## Assignment_Average

In [67]:
# Check Assignment_Average

print(
    "Missing values:",
    df_cleaned["Assignment_Average"].isna().sum()
)

invalid_assignment = df_cleaned[
    (df_cleaned["Assignment_Average"] < 0) |
    (df_cleaned["Assignment_Average"] > 100)
]

print(
    "Invalid Assignment_Average records:",
    len(invalid_assignment)
)

invalid_assignment[
    ["Student_ID", "Student_Name", "Assignment_Average"]
]

Missing values: 75
Invalid Assignment_Average records: 0


,Student_ID,Student_Name,Assignment_Average


## Cleaning Result: Assignment_Average

The `Assignment_Average` column was checked for values outside the valid range of 0–100.

No invalid values were found.

The column contains:
- 75 missing values
- 0 invalid values
- All non-missing values within the valid range

## Internal_Test_Marks

In [68]:
# Check Internal_Test_Marks

print(
    "Missing values:",
    df_cleaned["Internal_Test_Marks"].isna().sum()
)

invalid_internal = df_cleaned[
    (df_cleaned["Internal_Test_Marks"] < 0) |
    (df_cleaned["Internal_Test_Marks"] > 100)
]

print(
    "Invalid Internal_Test_Marks records:",
    len(invalid_internal)
)

invalid_internal[
    ["Student_ID", "Student_Name", "Internal_Test_Marks"]
]

Missing values: 37
Invalid Internal_Test_Marks records: 0


,Student_ID,Student_Name,Internal_Test_Marks


## Cleaning Result: Internal_Test_Marks

The `Internal_Test_Marks` column was checked for values outside the valid range of 0–100.

No invalid values were found.

The column contains:
- 37 missing values
- 0 invalid values
- All non-missing values within the valid range

## Practical_Marks

In [69]:
# Check Practical_Marks

print(
    "Missing values:",
    df_cleaned["Practical_Marks"].isna().sum()
)

invalid_practical = df_cleaned[
    (df_cleaned["Practical_Marks"] < 0) |
    (df_cleaned["Practical_Marks"] > 100)
]

print(
    "Invalid Practical_Marks records:",
    len(invalid_practical)
)

invalid_practical[
    ["Student_ID", "Student_Name", "Practical_Marks"]
]

Missing values: 0
Invalid Practical_Marks records: 0


,Student_ID,Student_Name,Practical_Marks


## Cleaning Result: Practical_Marks

The `Practical_Marks` column was checked for missing and invalid values.

No issues were found.

The column contains:
- 0 missing values
- 0 invalid values
- All values within the valid range of 0–100

## Participation_Score

In [70]:
# Check Participation_Score

print(
    "Missing values:",
    df_cleaned["Participation_Score"].isna().sum()
)

invalid_participation = df_cleaned[
    (df_cleaned["Participation_Score"] < 0) |
    (df_cleaned["Participation_Score"] > 100)
]

print(
    "Invalid Participation_Score records:",
    len(invalid_participation)
)

invalid_participation[
    ["Student_ID", "Student_Name", "Participation_Score"]
]

Missing values: 109
Invalid Participation_Score records: 0


,Student_ID,Student_Name,Participation_Score


## Cleaning Result: Participation_Score

The `Participation_Score` column was checked for missing and invalid values.

No invalid values were found.

The column contains:
- 109 missing values
- 0 invalid values
- All non-missing values within the valid range of 0–100

## Backlogs

In [71]:
# Check Backlogs

print(
    "Missing values:",
    df_cleaned["Backlogs"].isna().sum()
)

invalid_backlogs = df_cleaned[
    (df_cleaned["Backlogs"] < 0) |
    (df_cleaned["Backlogs"] > 20)
]

print(
    "Invalid Backlogs records:",
    len(invalid_backlogs)
)

invalid_backlogs[
    ["Student_ID", "Student_Name", "Backlogs"]
]

Missing values: 0
Invalid Backlogs records: 3


,Student_ID,Student_Name,Backlogs
287,S23CSE086,Sumit Kundu,-1
1519,S24CSE330,Kunal Banerjee,-1
2445,S25IT042,Vikram Nair,-1


In [72]:
print("\nBacklogs statistics:")
print(df_cleaned["Backlogs"].describe())

print("\nBacklog value counts:")
print(df_cleaned["Backlogs"].value_counts().sort_index())


Backlogs statistics:
count    2500.000000
mean        0.377200
std         0.745079
min        -1.000000
25%         0.000000
50%         0.000000
75%         1.000000
max         5.000000
Name: Backlogs, dtype: float64

Backlog value counts:
Backlogs
-1       3
 0    1815
 1     509
 2     114
 3      37
 4      12
 5      10
Name: count, dtype: int64


Step 1 — Replace invalid backlogs

In [73]:
# Replace invalid Backlogs values with NaN

invalid_backlogs_mask = df_cleaned["Backlogs"] < 0

df_cleaned.loc[
    invalid_backlogs_mask,
    "Backlogs"
] = np.nan

print(
    "Invalid Backlogs replaced with NaN:",
    invalid_backlogs_mask.sum()
)

Invalid Backlogs replaced with NaN: 3


Step 2 — Verify

In [74]:
print("Missing Backlogs:",
      df_cleaned["Backlogs"].isna().sum())

print("\nBacklogs statistics:")
print(df_cleaned["Backlogs"].describe())

print("\nBacklog value counts:")
print(df_cleaned["Backlogs"].value_counts().sort_index())

Missing Backlogs: 3

Backlogs statistics:
count    2497.000000
mean        0.378855
std         0.743994
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max         5.000000
Name: Backlogs, dtype: float64

Backlog value counts:
Backlogs
0.0    1815
1.0     509
2.0     114
3.0      37
4.0      12
5.0      10
Name: count, dtype: int64


## Cleaning Result: Backlogs

Three invalid backlog values of `-1` were identified.

Since backlog count cannot be negative, these values were replaced with `NaN`.

After cleaning:
- Valid records: 2,497
- Missing values: 3
- Minimum: 0
- Maximum: 5
- Median: 0

## Missing-Value Treatment

In [75]:
# Check missing values after data cleaning

missing_summary = pd.DataFrame({
    "Missing_Count": df_cleaned.isna().sum(),
    "Missing_Percentage": (
        df_cleaned.isna().sum() / len(df_cleaned) * 100
    ).round(2)
})

missing_summary = missing_summary[
    missing_summary["Missing_Count"] > 0
].sort_values(
    "Missing_Count",
    ascending=False
)

missing_summary

,Missing_Count,Missing_Percentage
Participation_Score,109,4.36
Attendance_Percentage,88,3.52
Sleep_Hours,79,3.16
Study_Resources,75,3.00
Assignment_Average,75,3.00
Study_Hours_Per_Day,65,2.60
Previous_Semester_Marks,50,2.00
Internal_Test_Marks,37,1.48
Internet_Access,25,1.00
Final_Exam_Marks,20,0.80


Step 1 — Fill numerical columns with median

In [76]:
# Numerical columns for median imputation

numeric_imputation_columns = [
    "Age",
    "Study_Hours_Per_Day",
    "Attendance_Percentage",
    "Previous_Semester_Marks",
    "Assignment_Average",
    "Internal_Test_Marks",
    "Practical_Marks",
    "Participation_Score",
    "Sleep_Hours",
    "Backlogs"
]

for column in numeric_imputation_columns:
    median_value = df_cleaned[column].median()

    df_cleaned[column] = df_cleaned[column].fillna(median_value)

    print(
        f"{column}: missing values filled with median = {median_value}"
    )

Age: missing values filled with median = 20.0
Study_Hours_Per_Day: missing values filled with median = 2.5
Attendance_Percentage: missing values filled with median = 77.0
Previous_Semester_Marks: missing values filled with median = 62.95
Assignment_Average: missing values filled with median = 13.0
Internal_Test_Marks: missing values filled with median = 19.5
Practical_Marks: missing values filled with median = 34.0
Participation_Score: missing values filled with median = 6.0
Sleep_Hours: missing values filled with median = 6.5
Backlogs: missing values filled with median = 0.0


Step 2 — Handle Study_Resources and Internet_Access

In [77]:
# Fill missing categorical values with the mode

categorical_imputation_columns = [
    "Internet_Access",
    "Study_Resources"
]

for column in categorical_imputation_columns:
    mode_value = df_cleaned[column].mode()[0]

    df_cleaned[column] = df_cleaned[column].fillna(mode_value)

    print(
        f"{column}: missing values filled with mode = {mode_value}"
    )

Internet_Access: missing values filled with mode = Yes
Study_Resources: missing values filled with mode = Both


Step 4 — Verify the remaining missing values

In [78]:
# Check remaining missing values

remaining_missing = pd.DataFrame({
    "Missing_Count": df_cleaned.isna().sum(),
    "Missing_Percentage": (
        df_cleaned.isna().sum() / len(df_cleaned) * 100
    ).round(2)
})

remaining_missing = remaining_missing[
    remaining_missing["Missing_Count"] > 0
].sort_values(
    "Missing_Count",
    ascending=False
)

remaining_missing

,Missing_Count,Missing_Percentage
Final_Exam_Marks,20,0.80
Pass_Fail,11,0.44
Performance_Category,11,0.44


Step 1 — Inspect the 20 missing final-exam marks

In [79]:
# Students with missing Final_Exam_Marks

missing_final = df_cleaned[
    df_cleaned["Final_Exam_Marks"].isna()
]

missing_final[
    [
        "Student_ID",
        "Student_Name",
        "Previous_Semester_Marks",
        "Assignment_Average",
        "Internal_Test_Marks",
        "Practical_Marks",
        "Pass_Fail",
        "Performance_Category"
    ]
]

,Student_ID,Student_Name,Previous_Semester_Marks,Assignment_Average,Internal_Test_Marks,Practical_Marks,Pass_Fail,Performance_Category
23,S23AIM023,Ari Khan,55.0,9.5,17.0,21,Pass,Needs Improvement
32,S23AIM032,Sakshi Rao,54.3,15.5,19.5,36,Pass,Average
441,S23ECE002,Tanmay Pal,81.0,17.5,30.0,42,Pass,Excellent
475,S23ECE035,Anirban Mondal,61.0,11.0,18.5,33,NaN,NaN
591,S23IT002,Isha Halder,64.6,12.0,22.0,35,NaN,NaN
672,S23IT083,Nisha Ghosh,89.0,20.0,27.5,44,NaN,NaN
695,S23IT106,Vishal Laha,77.0,14.5,20.0,32,NaN,NaN
804,S23ME045,Ritwik Ansari,53.4,14.0,15.0,33,Pass,Needs Improvement
969,S24AIM095,Ashish Shah,65.0,15.0,13.0,36,NaN,NaN
1101,S24AIM227,Rituparna Gupta,52.8,11.0,17.0,26,Pass,Good


Step 2 — Inspect the 11 missing target records

In [80]:
# Students with missing Pass_Fail or Performance_Category

missing_target = df_cleaned[
    df_cleaned["Pass_Fail"].isna() |
    df_cleaned["Performance_Category"].isna()
]

missing_target[
    [
        "Student_ID",
        "Student_Name",
        "Final_Exam_Marks",
        "Previous_Semester_Marks",
        "Assignment_Average",
        "Internal_Test_Marks",
        "Practical_Marks",
        "Pass_Fail",
        "Performance_Category"
    ]
]

,Student_ID,Student_Name,Final_Exam_Marks,Previous_Semester_Marks,Assignment_Average,Internal_Test_Marks,Practical_Marks,Pass_Fail,Performance_Category
475,S23ECE035,Anirban Mondal,NaN,61.0,11.0,18.5,33,NaN,NaN
591,S23IT002,Isha Halder,NaN,64.6,12.0,22.0,35,NaN,NaN
672,S23IT083,Nisha Ghosh,NaN,89.0,20.0,27.5,44,NaN,NaN
695,S23IT106,Vishal Laha,NaN,77.0,14.5,20.0,32,NaN,NaN
969,S24AIM095,Ashish Shah,NaN,65.0,15.0,13.0,36,NaN,NaN
1241,S24CSE053,Piyush Majumdar,NaN,72.9,13.5,21.5,35,NaN,NaN
1424,S24CSE236,Pooja Datta,NaN,73.0,20.0,24.5,36,NaN,NaN
1569,S24ECE016,Piyush Sinha,NaN,69.0,15.5,20.0,33,NaN,NaN
1731,S24ECE178,Arnab Majumdar,NaN,81.0,16.0,20.5,33,NaN,NaN
1865,S24IT094,Abhishek Gupta,NaN,69.0,15.5,23.0,43,NaN,NaN


Handle missing Final_Exam_Marks

In [81]:
# Remove students with missing Final_Exam_Marks

before = len(df_cleaned)

df_cleaned = df_cleaned.dropna(
    subset=["Final_Exam_Marks"]
).copy()

after = len(df_cleaned)

print("Rows before:", before)
print("Rows after:", after)
print("Rows removed:", before - after)

Rows before: 2500
Rows after: 2480
Rows removed: 20


In [82]:
print(
    "Missing Final_Exam_Marks:",
    df_cleaned["Final_Exam_Marks"].isna().sum()
)

Missing Final_Exam_Marks: 0


In [83]:
print("Missing Pass_Fail:",
      df_cleaned["Pass_Fail"].isna().sum())

print("Missing Performance_Category:",
      df_cleaned["Performance_Category"].isna().sum())

Missing Pass_Fail: 0
Missing Performance_Category: 0


In [84]:
# Check remaining missing values

missing_summary = pd.DataFrame({
    "Missing_Count": df_cleaned.isna().sum(),
    "Missing_Percentage": (
        df_cleaned.isna().sum() / len(df_cleaned) * 100
    ).round(2)
})

missing_summary = missing_summary[
    missing_summary["Missing_Count"] > 0
].sort_values(
    by="Missing_Count",
    ascending=False
)

missing_summary

,Missing_Count,Missing_Percentage


Verify everything

In [85]:
print("Total missing values:", df_cleaned.isna().sum().sum())

Total missing values: 0


In [86]:
print("Final dataset shape:", df_cleaned.shape)

Final dataset shape: (2480, 20)


# Save the cleaned dataset

In [87]:
df_cleaned.to_csv(
    "../data/processed/student_performance_cleaned.csv",
    index=False
)

print("Cleaned dataset saved successfully!")

Cleaned dataset saved successfully!


In [88]:
print("Shape:", df_cleaned.shape)
print("Missing values:", df_cleaned.isna().sum().sum())
print("Duplicate rows:", df_cleaned.duplicated().sum())
print("Duplicate Student IDs:", df_cleaned["Student_ID"].duplicated().sum())

Shape: (2480, 20)
Missing values: 0
Duplicate rows: 0
Duplicate Student IDs: 0


In [89]:
import os

print(os.path.exists("../data/processed/student_performance_cleaned.csv"))

True


In [90]:
df_check = pd.read_csv(file_path)

print("Saved dataset shape:", df_check.shape)
print("Missing values:", df_check.isna().sum().sum())
print("Duplicate rows:", df_check.duplicated().sum())
print("Duplicate Student IDs:", df_check["Student_ID"].duplicated().sum())

Saved dataset shape: (2540, 20)
Missing values: 622
Duplicate rows: 34
Duplicate Student IDs: 40


In [91]:
import os

file_path = "../data/processed/student_performance_cleaned.csv"

print("File exists:", os.path.exists(file_path))

File exists: True


In [92]:
# Final validation of cleaned dataset

print("Dataset shape:", df_cleaned.shape)
print("Total missing values:", df_cleaned.isna().sum().sum())
print("Duplicate rows:", df_cleaned.duplicated().sum())
print("Duplicate Student IDs:", df_cleaned["Student_ID"].duplicated().sum())

Dataset shape: (2480, 20)
Total missing values: 0
Duplicate rows: 0
Duplicate Student IDs: 0


In [93]:
df_check = pd.read_csv(file_path)

print("Saved dataset shape:", df_check.shape)
print("Missing values:", df_check.isna().sum().sum())
print("Duplicate rows:", df_check.duplicated().sum())
print("Duplicate Student IDs:", df_check["Student_ID"].duplicated().sum())

Saved dataset shape: (2480, 20)
Missing values: 0
Duplicate rows: 0
Duplicate Student IDs: 0


In [94]:
df_check.head()

,Student_ID,Student_Name,Age,Gender,Department,Semester,Study_Hours_Per_Day,Attendance_Percentage,Previous_Semester_Marks,Assignment_Average,Internal_Test_Marks,Practical_Marks,Participation_Score,Sleep_Hours,Internet_Access,Study_Resources,Backlogs,Final_Exam_Marks,Pass_Fail,Performance_Category
0,S23AIM001,Tanushree Tiwari,20.0,Female,CSE (AI & ML),5,3.50,89.5,76.5,19.0,25.0,40,9.0,7.0,Yes,Textbooks,0.0,77.0,Pass,Very Good
1,S23AIM002,Nitin Kundu,21.0,Male,CSE (AI & ML),5,1.00,71.5,32.0,8.0,4.0,22,7.0,7.0,Yes,Online Courses,4.0,16.0,Fail,Needs Improvement
2,S23AIM003,Amit Laha,21.0,Male,CSE (AI & ML),5,5.50,78.4,82.0,20.0,26.5,45,5.0,6.0,Yes,Notes Only,2.0,80.0,Pass,Very Good
3,S23AIM004,Akash Thakur,21.0,Male,CSE (AI & ML),6,4.00,78.9,67.7,16.5,21.5,38,6.0,5.0,Yes,Online Courses,0.0,63.0,Pass,Good
4,S23AIM005,Aarav Rao,22.0,Male,cse aiml,6,2.94,65.4,49.0,10.0,13.0,34,6.0,5.0,Yes,Both,0.0,47.0,Pass,Needs Improvement


In [97]:
department_mapping = {
    "CSE (AI & ML)": "CSE (AI & ML)",
    "CSE (AIML)": "CSE (AI & ML)",
    "AI & ML": "CSE (AI & ML)",
    "CSE(AIML)": "CSE (AI & ML)",
    "CSE-AIML": "CSE (AI & ML)",
    "cse aiml": "CSE (AI & ML)",

    "Computer Science & Engineering": "Computer Science & Engineering",
    "cse": "Computer Science & Engineering",
    "Comp Sci & Engg": "Computer Science & Engineering",
    "CSE": "Computer Science & Engineering",
    "Computer Science": "Computer Science & Engineering",
    "C.S.E": "Computer Science & Engineering",

    "Information Technology": "Information Technology",
    "Info Tech": "Information Technology",
    "information technology": "Information Technology",
    "IT": "Information Technology",
    "I.T.": "Information Technology",

    "Electronics & Communication": "Electronics & Communication",
    "E&C": "Electronics & Communication",
    "ece": "Electronics & Communication",
    "ECE": "Electronics & Communication",
    "Electronics & Comm.": "Electronics & Communication",

    "Mechanical Engineering": "Mechanical Engineering",
    "Mech. Engg": "Mechanical Engineering",
    "mech": "Mechanical Engineering",
    "Mechanical": "Mechanical Engineering",
    "ME": "Mechanical Engineering"
}

df_cleaned["Department"] = df_cleaned["Department"].replace(department_mapping)

In [100]:
print("Gender:")
print(df_cleaned["Gender"].value_counts())

print("\nInternet Access:")
print(df_cleaned["Internet_Access"].value_counts())

print("\nStudy Resources:")
print(df_cleaned["Study_Resources"].value_counts())

Gender:
Gender
Male       1330
Female      948
Other        32
M            27
Male         27
FEMALE       25
MALE         24
male         22
F            18
Female       16
female       11
Name: count, dtype: int64

Internet Access:
Internet_Access
Yes    2157
No      200
YES      53
Y        32
yes      28
no        4
NO        3
N         3
Name: count, dtype: int64

Study Resources:
Study_Resources
Both              859
Online Courses    636
Textbooks         520
Notes Only        371
online courses     19
both               15
Both               13
Online             13
Notes              10
Text Books          9
textbooks           8
notes only          7
Name: count, dtype: int64


## Standardizing Categorical Variables

Some categorical variables contain multiple representations of the same category.

For example, `Male`, `MALE`, `male`, and `M` represent the same gender category.

Similarly, Internet Access and Study Resources contain variations in capitalization, abbreviations, and spacing.

These values are standardized to consistent category names before machine learning preprocessing.

In [101]:
# Standardize Gender

gender_mapping = {
    "M": "Male",
    "MALE": "Male",
    "Male": "Male",
    "Male ": "Male",
    "male": "Male",
    "F": "Female",
    "FEMALE": "Female",
    "Female": "Female",
    "Female ": "Female",
    "female": "Female",
    "Other": "Other"
}

df_cleaned["Gender"] = df_cleaned["Gender"].replace(gender_mapping)


# Standardize Internet Access

internet_mapping = {
    "Y": "Yes",
    "YES": "Yes",
    "Yes": "Yes",
    "yes": "Yes",
    "N": "No",
    "NO": "No",
    "No": "No",
    "no": "No"
}

df_cleaned["Internet_Access"] = df_cleaned["Internet_Access"].replace(
    internet_mapping
)


# Standardize Study Resources

study_resources_mapping = {
    "Both": "Both",
    "Both ": "Both",
    "both": "Both",

    "Notes": "Notes Only",
    "Notes Only": "Notes Only",
    "notes only": "Notes Only",

    "Online": "Online Courses",
    "Online Courses": "Online Courses",
    "online courses": "Online Courses",

    "Text Books": "Textbooks",
    "Textbooks": "Textbooks",
    "textbooks": "Textbooks"
}

df_cleaned["Study_Resources"] = df_cleaned["Study_Resources"].replace(
    study_resources_mapping
)

print("Categorical variables standardized successfully!")

Categorical variables standardized successfully!


In [102]:
print("Gender:")
print(df_cleaned["Gender"].value_counts())

print("\nInternet Access:")
print(df_cleaned["Internet_Access"].value_counts())

print("\nStudy Resources:")
print(df_cleaned["Study_Resources"].value_counts())

Gender:
Gender
Male      1430
Female    1018
Other       32
Name: count, dtype: int64

Internet Access:
Internet_Access
Yes    2270
No      210
Name: count, dtype: int64

Study Resources:
Study_Resources
Both              887
Online Courses    668
Textbooks         537
Notes Only        388
Name: count, dtype: int64


In [103]:
df_cleaned.to_csv(
    "../data/processed/student_performance_cleaned.csv",
    index=False
)

print("Cleaned dataset saved successfully!")

Cleaned dataset saved successfully!


In [96]:
print(df.head())

  Student_ID      Student_Name  Age  Gender     Department  Semester  \
0  S23AIM001  Tanushree Tiwari   20  Female  CSE (AI & ML)         5   
1  S23AIM002       Nitin Kundu   21    Male  CSE (AI & ML)         5   
2  S23AIM003         Amit Laha   21    Male  CSE (AI & ML)         5   
3  S23AIM004      Akash Thakur   21    Male  CSE (AI & ML)         6   
4  S23AIM005         Aarav Rao   22    Male       cse aiml         6   

  Study_Hours_Per_Day Attendance_Percentage  Previous_Semester_Marks  \
0                 3.5                  89.5                     76.5   
1                   1                  71.5                     32.0   
2                 5.5                  78.4                     82.0   
3                   4                  78.9                     67.7   
4                2.94                  65.4                     49.0   

   Assignment_Average  Internal_Test_Marks  Practical_Marks  \
0                19.0                 25.0               40   
1       

In [98]:
print(df_cleaned["Department"].value_counts())

Department
Computer Science & Engineering    691
CSE (AI & ML)                     586
Information Technology            445
Electronics & Communication       424
Mechanical Engineering            334
Name: count, dtype: int64


In [99]:
df_cleaned.to_csv(
    "../data/processed/student_performance_cleaned.csv",
    index=False
)

print("Cleaned dataset saved successfully!")

Cleaned dataset saved successfully!
